### ISSUE TITLE
[Feature Request] Prometheus metrics from kopia server

### ISSUE THREAD

**Issue by hickinbottoms** [kopia/kopia#609]
It would be great if kopia server exported Prometheus-format metrics, e.g. under "/metrics", for things such as:

- Snapshot age and size
- Maintenance interval for quick/full, plus time until the next scheduled run
- Recent maintenance results

That would allow monitoring via a dashboard and, more importantly, raising alerts if the time since the last snapshot grew larger than expected.

Probably that URL would be protected via the same user authentication as access to the web UI. If more advanced user management is implemented (#548), then maybe a "metrics" permission could be defined.

If this seems like a good idea I could suggest how those metrics could look.

**Comment by jkowalski** [comment]
Kopia has OpenCensus integration for low-level repository operations (basic performance counters). Currently `/metrics` served by `kopia server` is excluded from authentication, but we can change that.

The metrics you're proposing sound great. 

Along similar lines I was also thinking of adding some kind of textual report where compute these metrics and send them in a nicely-formatted email, say once a week or when something goes bad. This would work from a running server or could be generated on-demand e.g. from cron (`kopia report`) 

Another idea - we could save those reports over time in the repository itself (treating them effectively as snapshots, where they would be subject to retention, garbage collection, etc.). Perhaps maintenance logs could also be treated

**Comment by hickinbottoms** [comment]
I'd not noticed the /metrics endpoint was already there, so that's a nice surprise.

I'll spend a little while thinking of the metric detail for my suggestion above then I'll post an example here.

**Comment by hickinbottoms** [comment]
Here’s an example of what the returned metrics could look like.

Snapshot info:

```text
# HELP kopia_snapshot_count Number of completed snapshots of a path
# TYPE kopia_snapshot_count gauge
kopia_snapshot_count{owner="user@host",path="/some/path/1",root="kfe997567fb1cf8a13341e4ca11652f70"} 4
kopia_snapshot_count{owner="user@host",path="/some/path/2",root="kb0fe36a289c8fca8e1f2d6905bf06f9e"} 3
kopia_snapshot_count{owner="user@host",path="/some/path/3",root="k6b073a118aac7fdf3c46bb21fc03ab1e"} 6
# HELP kopia_last_snapshot_age_seconds Time since most recent completed Kopia snapshot
# TYPE kopia_last_snapshot_age_seconds gauge
kopia_last_snapshot_age_seconds{owner="user@host",path="/some/path/1",root="kfe997567fb1cf8a13341e4ca11652f70"} 9000
kopia_last_snapshot_age_seconds{owner="user@host",path="/some/path/2",root="kb0fe36a289c8fca8e1f2d6905bf06f9e"} 9005
kopia_last_snapshot_age_seconds{owner="user@host",path="/some/path/3",root="k6b073a118aac7fdf3c46bb21fc03ab1e"} 9010
# HELP kopia_last_snapshot_size_bytes Size of most recent completed snapshot
# TYPE kopia_last_snapshot_size_bytes gauge
kopia_last_snapshot_size_bytes{owner="user@host",path="/some/path/1",root="kfe997567fb1cf8a13341e4ca11652f70"} 2097152
kopia_last_snapshot_size_bytes{owner="user@host",path="/some/path/2",root="kb0fe36a289c8fca8e1f2d6905bf06f9e"} 2096152
kopia_last_snapshot_size_bytes{owner="user@host",path="/some/path/3",root="k6b073a118aac7fdf3c46bb21fc03ab1e"} 2095152
# HELP kopia_last_snapshot_file_count Number of files in the most recent completed snapshot
# TYPE kopia_last_snapshot_file_count gauge
kopia_last_snapshot_file_count{owner="user@host",path="/some/path/1",root="kfe997567fb1cf8a13341e4ca11652f70"} 10923
kopia_last_snapshot_file_count{owner="user@host",path="/some/path/2",root="kb0fe36a289c8fca8e1f2d6905bf06f9e"} 345
kopia_last_snapshot_file_count{owner="user@host",path="/some/path/3",root="k6b073a118aac7fdf3c46bb21fc03ab1e"} 987
# HELP kopia_last_snapshot_directory_count Number of directories in the most recent completed snapshot
# TYPE kopia_last_snapshot_directory_count gauge
kopia_last_snapshot_directory_count{owner="user@host",path="/some/path/1",root="kfe997567fb1cf8a13341e4ca11652f70"} 2432
kopia_last_snapshot_directory_count{owner="user@host",path="/some/path/2",root="kb0fe36a289c8fca8e1f2d6905bf06f9e"} 42
kopia_last_snapshot_directory_count{owner="user@host",path="/some/path/3",root="k6b073a118aac7fdf3c46bb21fc03ab1e"} 75
```

If there are no completed snapshots then the following aren’t returned at all; incomplete snapshots don’t count.

-   `kopia_last_*`
-   `kopia_snapshot_count`

Maintenance info:

```text
# HELP kopia_maintenance_interval_seconds Time between scheduled maintenance tasks
# TYPE kopia_maintenance_interval_seconds gauge
kopia_maintenance_interval_seconds{owner="user@host",type="quick"} 3600
kopia_maintenance_interval_seconds{owner="user@host",type="full"} 86400
# HELP kopia_maintenance_next_run_seconds Time until next maintenance task run
# TYPE kopia_maintenance_next_run_seconds gauge
kopia_maintenance_next_run_seconds{owner="user@host",type="quick"} 2321
kopia_maintenance_next_run_seconds{owner="user@host",type="full"} 67127
# HELP kopia_maintenance_last_run_seconds Duration of last maintenance task
# TYPE kopia_maintenance_last_run_seconds gauge
kopia_maintenance_last_run_seconds{owner="user@host",task="full-delete-blobs"} 95
kopia_maintenance_last_run_seconds{owner="user@host",task="full-drop-deleted-content"} 27
kopia_maintenance_last_run_seconds{owner="user@host",task="full-rewrite-contents"} 476
kopia_maintenance_last_run_seconds{owner="user@host",task="index-compaction"} 0
kopia_maintenance_last_run_seconds{owner="user@host",task="quick-delete-blobs"} 0
kopia_maintenance_last_run_seconds{owner="user@host",task="quick-rewrite-contents"} 2
kopia_maintenance_last_run_seconds{owner="user@host",task="snapshot-gc"} 68
```

If either quick or full maintenance is not set then the following are not returned for either or both types:

-   `kopia_maintenance_interval_seconds`
-   `kopia_maintenance_next_run_seconds`
-   `kopia_maintenance_last_run_seconds`

Prometheus will scrape these metrics frequently (e.g. every 15 seconds), so if they take a non-trivial amount of time to derive then maybe cache them for a period and only re-generate them if they’re stale for more than a defined period (e.g. 1 minute, or 5 minutes?). Maybe that caching isn’t needed because the UI is pretty quick.

**Comment by altitude1326** [comment]
Hi @jkowalski , first, thanks a lot for this great software !
I have a question about prometheus. I see is source code this : 
https://github.com/kopia/kopia/blob/991c08e4b464eab530fcd0e2b408c8b546719a87/cli/app.go#L227
so i start kopia server like this to test : 
`kopia server start --address http://0.0.0.0:51515 --random-password --insecure --server-username=moi --metrics-listen-addr http://0.0.0.0:51516`
I well see in logs : 
`starting prometheus metrics on http://0.0.0.0:51516`
But in fact server not started on port 51516. I missed something ?
King regards,
Antoine.

**Comment by barnumbirr** [comment]
Hi,

same as altitude1326, running the following:

```bash
$ kopia server start --tls-cert-file /opt/Kopia/kopia.cert --tls-key-file /opt/Kopia/kopia.key --address 0.0.0.0:51515 --metrics-lis
ten-addr 0.0.0.0:51516
starting prometheus metrics on 0.0.0.0:51516

Server will allow connections from users whose accounts are stored in the repository.
User accounts can be added using 'kopia server user add'.

SERVER ADDRESS: https://[::]:51515
Open the address above in a web browser to use the UI.
```

results in the following:

```bash
$ curl -v https://10.0.1.100:51516/metrics
*   Trying 10.0.1.100:51516...
* connect to 10.0.1.100 port 51516 failed: Connection refused
* Failed to connect to 10.0.1.100 port 51516 after 0 ms: Connection refused
* Closing connection 0
curl: (7) Failed to connect to 10.0.1.100 port 51516 after 0 ms: Connection refused
```

Is this feature broken?

Thanks.

**Comment by sparfenov** [comment]
Vote for @hickinbottoms proposal!
It would be great

**Comment by electrofloat** [comment]
Hmm... why was this closed as not planned?
Reading through this, there was already a /metrics endpoint but was not working?
And the other suggested metrics are not planned at all?

**Comment by frli4797** [comment]
> Hi,
> 
> same as altitude1326, running the following:
> 
> ```shell
> $ kopia server start --tls-cert-file /opt/Kopia/kopia.cert --tls-key-file /opt/Kopia/kopia.key --address 0.0.0.0:51515 --metrics-lis
> ten-addr 0.0.0.0:51516
> starting prometheus metrics on 0.0.0.0:51516
> 
> Server will allow connections from users whose accounts are stored in the repository.
> User accounts can be added using 'kopia server user add'.
> 
> SERVER ADDRESS: https://[::]:51515
> Open the address above in a web browser to use the UI.
> ```
> 
> results in the following:
> 
> ```shell
> $ curl -v https://10.0.1.100:51516/metrics
> *   Trying 10.0.1.100:51516...
> * connect to 10.0.1.100 port 51516 failed: Connection refused
> * Failed to connect to 10.0.1.100 port 51516 after 0 ms: Connection refused
> * Closing connection 0
> curl: (7) Failed to connect to 10.0.1.100 port 51516 after 0 ms: Connection refused
> ```
> 
> Is this feature broken?
> 
> Thanks.
@altitude1326 @barnumbirr I managed to use Prometheus pushgw to get some metrics going as described here. #3549. Might be an option for you as well even though I receive things on a Victoriametrics instance. However, this should be working with your usual Prometheus pushgw as well. 

Still, there is no reporting of the latest `kopia server snapshot create` or similar, that remains to be implemented by someone(tm). I'd love to take a crack at it, but I have literally zero experience with Go. Have a hard time seeing any PR of mine being merged. :)

**Comment by alvistar** [comment]
Hi,
I wrote simple exporter in Python.

https://github.com/alvistar/kopia-exporter

This would be running "kopia snapshot list -n 1 --json" in intervals.

I would like as well to be integrated in main kopia to easy monitor backup statuses. I checked the code, but I am not proficient in Go and I think it will require a quite amount of time from my side.

If somebody wants to help and let me know where should the relevant parts of code to patch let me know, but not promises here.

Thanks for all the job.
